libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.cpp

```diff
@@ -0,0 +1,157 @@
+    this->hash_values = {
+        example.beta,
+        example.gamma,
+        example.alpha,
+        example.zeta,
+        example.nu,
+        example.u,
+    };
+}
+
+void bls12_381_test_vector_transcript_hasher::buffer_clear()
+{
+    this->buffer.clear();
+}
+
+size_t bls12_381_test_vector_transcript_hasher::buffer_size()
+{
+    return this->buffer.size();
+}
+
+void bls12_381_test_vector_transcript_hasher::add_element(
+    const libff::Fr<libff::bls12_381_pp> &element)
+{
+    // convert the Fr element into a string
+    std::string str;
+    {
+        std::ostringstream ss;
+        libff::field_write<libff::encoding_binary, libff::form_plain>(
+            element, ss);
+        str = ss.str();
+    }
+    // copy the string as a sequence of uint8_t elements at the end of
+    // the buffer
+    std::copy(str.begin(), str.end(), std::back_inserter(this->buffer));
+}
+
+void bls12_381_test_vector_transcript_hasher::add_element(
+    const libff::G1<libff::bls12_381_pp> &element)
+{
+    libff::G1<libff::bls12_381_pp> element_aff(element);
+    element_aff.to_affine_coordinates();
+
+    // convert the affine coordinates of the curve point into a string
+    std::string str;
+    {
+        std::ostringstream ss;
+        libff::group_write<
+            libff::encoding_binary,
+            libff::form_plain,
+            libff::compression_off>(element_aff, ss);
+        str = ss.str();
+    }
+    // copy the string as a sequence of uint8_t elements at the end of
+    // the buffer
+    std::copy(str.begin(), str.end(), std::back_inserter(this->buffer));
+}
+
+void bls12_381_test_vector_transcript_hasher::add_element(
+    const libff::G2<libff::bls12_381_pp> &element)
+{
+    libff::G2<libff::bls12_381_pp> element_aff(element);
+    element_aff.to_affine_coordinates();
+
+    // convert the affine coordinates of the curve point into a string
+    std::string str;
+    {
+        std::ostringstream ss;
+        libff::group_write<
+            libff::encoding_binary,
+            libff::form_plain,
+            libff::compression_off>(element_aff, ss);
+        str = ss.str();
+    }
+    // copy the string as a sequence of uint8_t elements at the end of
+    // the buffer
+    std::copy(str.begin(), str.end(), std::back_inserter(this->buffer));
+}
+
+libff::Fr<libff::bls12_381_pp> bls12_381_test_vector_transcript_hasher::
+    get_hash()
+{
+    size_t buffer_len = this->buffer.size();
+
+    // vector of valid lengths (\attention specific to BLS12-381)
+    const std::vector<size_t> length{288, 320, 416, 704, 896, 1120};
+
+    // If we are here, then the hasher buffer has invalid length so throw an
+    // exception
+    bool b_valid_length =
+        (0 != std::count(length.begin(), length.end(), buffer_len));
+    if (!b_valid_length) {
+        throw std::logic_error(
+            "Error: invalid length of transcript hasher buffer");
+    }
+
+    libff::Fr<libff::bls12_381_pp> challenge = 0;
```

Comment:
Not used until it is assigned below, at which point it is immediately returned.  Either remove (and just return `hash_values[i]`), or declare when it is assigned, so it can be `const`.

Reply:
Done.